### PR TITLE
[Outlook] Add sample to autotag external recipients when composing messages

### PR DESCRIPTION
**Pull request by ElizabethSamuel-MSFT**
|        Q        |                    A                    |
| --------------- | --------------------------------------- |
| Bug fix?        | no - yes? no                              |
| New feature?    | no - yes? yes                              |
| New sample?     | no - yes? yes                              |
| Related issues? | no |

## What's in this Pull Request?

New sample that demonstrates how to handle OnMessageRecipientsChanged event (now in preview). SessionData object (in preview) is also used, along with the item.body.appendOnSendAsync API. The sample tags an email subject as external if there are any external recipients.

Once this PR is merged, the manifest files will be updated to point to the main OfficeDev repo (not the forked version which is just for review in this PR for now.)

### CODE REVIEW

File: Samples/outlook-tag-external/src/commands/commands.js [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      }
+    });
+}
+/**
+ * Determines if there are any external recipients in the Cc field.
+ * If there are, updates the subject of the Outlook message
+ * and appends a disclaimer to the message body.
+ */
+function checkForExternalCc() {
+  console.log("tagExternal method"); //debugging
```
Comment (davidchesnut) [inline Samples/outlook-tag-external/src/commands/commands.js:63]:
```suggestion
  console.log("checkForExternalCc method"); //debugging
```

**Review by davidchesnut — CHANGES_REQUESTED** [review]
Looks good. Added a few notes. Thanks!

File: Samples/outlook-tag-external/README.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+### Handle the OnMessageRecipientsChanged event, manage session data, and call the appendOnSendAsync API
+
+When the user composes a message (including replies and forwards) and changes any recipients, Outlook will load the files specified in the manifest to handle the `OnMessageRecipientsChanged` event. Outlook on the web loads the **commands.html** page, which then also loads **commands.js**. In Outlook on Windows, **commands.js** is loaded directly but **commands.html** is not loaded.
+
+The **commands.js** file contains the `tagExternal_onMessageRecipientsChangedHandler` function that handles the `OnMessageRecipientsChanged` event from Outlook.
+
+Also, the **commands.js** file contains the following helper functions.
+
+- `checkForExternalTo`: Determines if there are any external users in the **To** field then sets a [SessionData](https://docs.microsoft.com/javascript/api/outlook/office.messagecompose?view=outlook-js-preview#sessionData) key named **tagExternalTo**.
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/README.md:161]: Should these be relative links?

File: Samples/outlook-tag-external/manifest.xml [hunk 3]
```diff
@@ -0,0 +1,88 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<OfficeApp xmlns="http://schemas.microsoft.com/office/appforoffice/1.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bt="http://schemas.microsoft.com/office/officeappbasictypes/1.0" xmlns:mailappor="http://schemas.microsoft.com/office/mailappversionoverrides/1.0" xsi:type="MailApp">
+  <Id>84cc5da3-2a5e-4373-8130-7c38a2a6eaf3</Id>
+  <Version>1.0.0.0</Version>
+  <ProviderName>Microsoft</ProviderName>
+  <DefaultLocale>en-US</DefaultLocale>
+  <DisplayName DefaultValue="PnP Tag External Recipients sample"/>
+  <Description DefaultValue="Updates the subject and adds a disclaimer if external recipients are included during message compose."/>
+  <IconUrl DefaultValue="https://elizabethsamuel-msft.github.io/PnP-OfficeAddins/Samples/outlook-tag-external/assets/icon-64.png"/>
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/manifest.xml:9]: Please move the icons into a directory in the PnP-OfficeAddins repo and link to them there.

File: Samples/outlook-tag-external/manifest.xml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      <Description resid="residAppDesc" />
+      <Requirements>
+        <bt:Sets DefaultMinVersion="1.3">
+          <bt:Set Name="Mailbox" />
+        </bt:Sets>
+      </Requirements>
+      <Hosts>
+        <Host xsi:type="MailHost">
+          <!-- Event-based activation happens in a lightweight runtime.-->
+          <!-- For more information, see https://docs.microsoft.com/office/dev/add-ins/outlook/autolaunch -->
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/manifest.xml:47]: Should this be a relative link?

File: Samples/outlook-tag-external/manifest.xml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+          <Runtimes>
+            <!-- HTML file including reference to or inline JavaScript event handlers.
+                This is used by Outlook on the web. -->
+            <Runtime resid="WebViewRuntime.Url">
+              <!-- JavaScript file containing event handlers. This is used by Outlook Desktop. -->
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/manifest.xml:52]:
```suggestion
              <!-- JavaScript file containing event handlers. This is used by Outlook on Windows. -->
```

File: Samples/outlook-tag-external/src/commands/commands.js [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        console.log("At least one recipients field includes external users"); //debugging
+        _tagExternal(true);
+      } else {
+        console.log("There are no external recipients"); //debugging
+        _tagExternal(false);
+      }
+    });
+}
+/**
+ * If there are any external recipients, tags the subject of the Outlook item
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/src/commands/commands.js:172]:
```suggestion
 * If there are any external recipients, prepends the subject of the Outlook item
```

File: Samples/outlook-tag-external/src/commands/commands.js [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+ * as "External" and appends a disclaimer to the item body. If there are
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/src/commands/commands.js:173]:
```suggestion
 * with "[External]" and appends a disclaimer to the item body. If there are
```

File: Samples/outlook-tag-external/manifest-localhost.xml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      </Requirements>
+      <Hosts>
+        <Host xsi:type="MailHost">
+          <!-- Event-based activation happens in a lightweight runtime.-->
+          <!-- For more information, see https://docs.microsoft.com/office/dev/add-ins/outlook/autolaunch -->
+          <Runtimes>
+            <!-- HTML file including reference to or inline JavaScript event handlers.
+                This is used by Outlook on the web. -->
+            <Runtime resid="WebViewRuntime.Url">
+              <!-- JavaScript file containing event handlers. This is used by Outlook Desktop. -->
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/manifest-localhost.xml:52]:
```suggestion
              <!-- JavaScript file containing event handlers. This is used by Outlook on Windows. -->
```

File: Samples/outlook-tag-external/manifest-localhost.xml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      </Hosts>
+      <Resources>
+        <bt:Images>
+          <bt:Image id="Icon.16x16" DefaultValue="https://localhost:3000/assets/icon-16.png"/>
+          <bt:Image id="Icon.32x32" DefaultValue="https://localhost:3000/assets/icon-32.png"/>
+          <bt:Image id="Icon.80x80" DefaultValue="https://localhost:3000/assets/icon-80.png"/>
+        </bt:Images>
+        <bt:Urls>
+          <bt:Url id="WebViewRuntime.Url" DefaultValue="https://localhost:3000/src/commands/commands.html" />
+          <!-- Entry needed for Outlook Desktop. -->
```
Comment (lindalu-MSFT) [inline Samples/outlook-tag-external/manifest-localhost.xml:76]:
```suggestion
          <!-- Entry needed for Outlook on Windows. -->
```

**Review by lindalu-MSFT — CHANGES_REQUESTED** [review]

File: Samples/outlook-tag-external/README.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+## Run the sample
+
+You can run this sample in Outlook on Windows or in a browser. The add-in web files are served from this repo on GitHub.
+
+1. Download the **manifest.xml** file from this sample to a folder on your computer.
+1. Sideload the add-in manifest in Outlook on Windows, or Outlook on the web by following the manual instructions in the article [Sideload Outlook add-ins for testing](https://docs.microsoft.com/office/dev/add-ins/outlook/sideload-outlook-add-ins-for-testing).
+
+### Try it out
+
+Once the add-in is loaded use the following steps to try out the functionality.
```
[the remark this answers is not in the document]
Comment (ElizabethSamuel-MSFT) [reply Samples/outlook-tag-external/README.md:76]: @lindalu-MSFT , per the [style guide](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/f/following), I thought this case should end with a period.

Here and other instances.

**Review by ElizabethSamuel-MSFT — COMMENTED** [review]
@lindalu-MSFT I won't be able to generate the OfficeDev links until these changes are committed to main. I'll make sure to do a subsequent PR with those updates.

File: Samples/outlook-tag-external/README.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+If you prefer to host the web server for the sample on your computer, follow these steps:
+
+1. You need http-server to run the local web server. If you haven't installed this yet, you can do this with the following command.
+
+    ```console
+    npm install --global http-server 
+    ```
+
+1. Use a tool such as openssl to generate a self-signed certificate that you can use for the web server. Move the cert.pem and key.pem files to the root folder for this sample.
+1. From a command prompt, go to the root folder and run the following command.
```
[the remark this answers is not in the document]
Comment (ElizabethSamuel-MSFT) [reply Samples/outlook-tag-external/README.md:99]: I'll make sure to update the guidance for the section with the applicable fixes from this edit pass.

File: Samples/outlook-tag-external/README.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+The manifest configures a runtime that is loaded specifically to handle event-based activation. The following `<Runtime>` element specifies an HTML page resource ID that loads the runtime in Outlook on the web. The `<Override>` element specifies the JavaScript file to load the runtime for Outlook on Windows because Outlook on Windows doesn't use the HTML page to load the runtime.
+
+```xml
+<Runtime resid="WebViewRuntime.Url">
+  <Override type="javascript" resid="JSRuntime.Url"/>
+</Runtime>
+...
+<bt:Url id="WebViewRuntime.Url" DefaultValue="https://elizabethsamuel-msft.github.io/PnP-OfficeAddins/Samples/outlook-tag-external/src/commands.html" />
+<bt:Url id="JSRuntime.Url" DefaultValue="https://elizabethsamuel-msft.github.io/PnP-OfficeAddins/Samples/outlook-tag-external/src/commands/commands.js" />
```
[the remark this answers is not in the document]
Comment (ElizabethSamuel-MSFT) [reply Samples/outlook-tag-external/README.md:125]: @lindalu-MSFT Yes, for now. Once this change is merged and the pages are generated, I'll update these links in a subsequent PR.

File: Samples/outlook-tag-external/manifest.xml (hunk 3 above)
Comment (ElizabethSamuel-MSFT) [reply Samples/outlook-tag-external/manifest.xml:9]: This is correct for now. Once this change is merged and the pages are generated, I'll update these links in a subsequent PR.

File: Samples/outlook-tag-external/README.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+<ExtendedPermission>AppendOnSend</ExtendedPermission>
+```
+
+### Handle the OnMessageRecipientsChanged event, manage session data, and call the appendOnSendAsync API
+
+When the user composes a message (including replies and forwards) and changes any recipients, Outlook will load the files specified in the manifest to handle the `OnMessageRecipientsChanged` event. Outlook on the web loads the **commands.html** page, which then also loads **commands.js**. In Outlook on Windows, **commands.js** is loaded directly but **commands.html** is not loaded.
+
+The **commands.js** file contains the `tagExternal_onMessageRecipientsChangedHandler` function that handles the `OnMessageRecipientsChanged` event from Outlook.
+
+Also, the **commands.js** file contains the following helper functions.
```
[the remark this answers is not in the document]
Comment (ElizabethSamuel-MSFT) [reply Samples/outlook-tag-external/README.md:152]: @lindalu-MSFT per the [style guide](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/f/following), I thought this case (and others similar in this PR) should end with a period.

[the remark this answers is not in the document]
Comment (lindalu-MSFT) [reply Samples/outlook-tag-external/README.md:152]: Totally up to you, as long as it's consistent throughout!

**Pull request merged**